### DISCUSSION TITLE
Method return type inference

### DISCUSSION THREAD

**Discussion by a11delavar** [dotnet/csharplang#6452] [Language Ideas] [answered]
Inferring method return types is one of the early issues [raised in the Roslyn repository](https://github.com/dotnet/roslyn/issues/17) which was rejected because of the technical implications at the time, and that [C# wasn't really to go down a heavy inference path](https://github.com/dotnet/roslyn/issues/17#issuecomment-120680802).

It's been 7 years and we've got fantastic related features such as [inferred lambda types](https://github.com/dotnet/roslyn/issues/14), which 
 [wasn't supposed to land](https://github.com/dotnet/roslyn/issues/3912#issue-94526628) at the time, and already laid the foundation to solve many of those technical issues.

So I opened this discussion to see whether there has been any follow-up discussion on the subject and what the stance of the C# team is in 2022.

**Comment by CyrusNajmabadi** [comment, marked as answer]
The stance is the same.  We will not do this.  The technical issues are hte same as before. 

To reiterate the issue:
Inferred return types completely break two things:

1. the invariant that top level signatures are always explicit.  This is a valuable aspect of the language that we do nto want to diminish.
2. the invariant that a method body edit is not visible outside of hte method.  this is heavily part of hte entire perf story of roslyn as it allows us to just 'recompile/reanalyze' a method body in isolation when an edit is made in it.  Taking this away will defeat one of (if not the top) perf optimizations we have for the entire roslyn stack.

**Reply by a11delavar** [reply]
Thanks for the answer!

> The technical issues are hte same as before.

Honestly the reasons outlined here are new to me who only followed the discussions in [roslyn#17](https://github.com/dotnet/roslyn/issues/17) (and [csharplang#2996](https://github.com/dotnet/csharplang/issues/2996)) back in the day.
Most of the discussion was about the implementation details half of which was about how the inference should really work for the compiler to have a solid level of certainty and prevent doing any guess-work, and it all is dealt with in C# 10 with inferred lambda types. Other points were about generics and circular references, etc.

> the invariant that top level signatures are always explicit. This is a valuable aspect of the language that we do nto want to diminish.

and the reason would be (among other things) your second point?

> the invariant that a method body edit is not visible outside of hte method. this is heavily part of hte entire perf story of roslyn as it allows us to just 'recompile/reanalyze' a method body in isolation when an edit is made in it. Taking this away will defeat one of (if not the top) perf optimizations we have for the entire roslyn stack.

Interesting. I was aware of extra work the compiler has to do to infer the return type of a given function, and possibly continue working up to its callers to reanalyze those and so on, but wouldn't have though it would hit the performance so badly. Good to know!

**Reply by HaloFour** [reply]
I think the challenges with inferring lambda types were solved by reducing where you could apply that inference.  For instance, you can't have a natural lambda type for a lambda that has inferred parameter types.  That is only allowed if the target delegate type is known:

```cs
Func<int, int, int> add1 = (x, y) => x + y; // legal
var add2 = (int x, int y) => x + y; // legal
var add3 = (x, y) => x + y; // error CS8917: The delegate type could not be inferred.
``` 

Because the parameter types must be known the signature of the lambda becomes much easier to infer.  The scoping reduces the amount of code that would have to be analyzed, and the lexical ordering of the assignments to inferred local effectively eliminates circular references.

**Reply by a11delavar** [reply]
@HaloFour 

> I think the challenges with inferring lambda types were solved by reducing where you could apply that inference. 

Interesting!

> Because the parameter types must be known the signature of the lambda becomes much easier to infer. The scoping reduces the amount of code that would have to be analyzed, and the lexical ordering of the assignments to inferred local effectively eliminates circular references.

For what it's worth, I must clarify that the inference of "all things" or even "so many variations like for lambdas" isn't meant here.

I only meant given "explicitly written parameter types" like before, the "return type" could be inferred by the compiler in the lowering process, though I'm not familiar with Roslyn and internal constrains due to performance.

So the return type would be optional.
```cs
/* void */ Write(string text) => Console.Write(text)
```

Now, let's put together something more realistic to show the better readability and reduction in boilerplate:
```cs
private async /* Task<Dictionary<Guid, IEnumerable<Member>>> */ GetMembersByReservationGuidsForHallAsync(Guid hallGuid) {
    var reservationGuids = GetReservationGuids(hallGuid); // Task<IEnumerable<Guid>>
    var members = GetAllMembers(hallGuid); // Task<IEnumerable<Member>>
    await Task.WhenAll(reservationGuids, members);
    
    // The compiler already infers the type of dictionary, and writes it out in the lowering process
    var dictionary = reservationGuids.Result
        .Select(guid => (guid, members.Result.Select(m => m.ReservationGuid == guid)))
        .ToDictionary(x => x.Item1, x => x.Item2);

    // Well the inferred type is also the return type of this function here.
    return dictionary;
}
```

**Reply by HaloFour** [reply]
IMO the contract is not boilerplate, it's much more important than what the method body actually does.  You now have to scan the entire method body to understand what the method might return, and if the method has multiple paths to return a value that can become very complicated.  Additionally, `async` does not imply `Task`, it's the type itself that determines how the compiler emits the internals of the asynchronous state machine, so you immediately end up with a circular inference situation.

Yes, the compiler can figure this out, but the compiler is not the most important parser that needs to quickly make sense of that code, and that code will be read significantly more frequently than it is written so the language should optimize for clarity over brevity.

**Reply by corruptedlotus** [reply]
But we have already been doing this without scanning the method body, in languages like TS. The implication of `Task` is neither a big deal, as it is always possible for the developer to explicitly specify the return type, `void` or whatever else. The visibility of return types has never been the full clarity for the one willing to know the actual **thing** that is returned by a method, in fact it has been far inferior to picking the correct naming and having a clean method body. Yet, inference doesn't remove the visibility anyway, compiler and LS are the ones doing the inferring and they can tell you, just like when you hover a `var` and the inferred type is crystal clear.

What I can see as mind boggling that you also mentioned is the problem of multiple-contracts-possible especially when it comes to tasks. Inference doesn't take away the ability to explicitly specify the return type, and, there's almost always the one-fits-all selection when it comes to such multi-choice inferences, and it's safe to say someone who hasn't explicitly specified and override to that behaviour is looking exactly for such selection.

**Reply by CyrusNajmabadi** [reply]
> But we have already been doing this without scanning the method body, in languages like TS.

And it's a very big deal in TS as well and the source of a lot of performance issues.

**Reply by HaloFour** [reply]
Several of the C# language members have also worked on TS.  The two languages have extremely different paradigms and goals.  Even if inference of this nature wouldn't drastically impact the performance of the compiler it's extremely unlikely such a change would be considered given that the team believes that it is important to be explicit about the contract, and that extends beyond just the types of the signature but also to nullability, requiredness, etc.

**Reply by a11delavar** [reply]
@HaloFour 

> IMO the contract is not boilerplate

I agree that in many cases it is not boilerplate and it will be explicit, but in other cases like writing a script or some UI code it is extremely helpful not to think about the contract of some private helper functions which of course do have their contract, but is not that relevant to the developer.

> You now have to scan the entire method body to understand what the method might return, and if the method has multiple paths to return a value that can become very complicated.

Having `var` in C# doesn't make us devs to reanalyze the whole code to understand the variable's types but rather you hover on the variable to see what the compiler infers to be its type in your code editor, which happens not often for me at least.

It would be the same for the methods' return types. Many language servers already do that, and I have first-hand experience in large projects of TypeScript and in many projects more that half of the code is inferred and the readability has not been a problem, in fact the opposite, it is extremely helpful that you don't want to repeat return types for some helper/proxy methods.
Furthermore editors like VS Code support additional UI tools like "Inlay Hints" to facilitate the DX, also not a bad idea for other inferences such as `var` itself, but it is rather nice-to-have.
![image](https://user-images.githubusercontent.com/38350963/190924653-a1c74359-38cf-4187-8c99-d91c0cd38c46.png)
![image](https://user-images.githubusercontent.com/38350963/190924656-945e0259-51df-4913-a912-aa012cdfe2b5.png)
![image](https://user-images.githubusercontent.com/38350963/190924659-7618c58c-514d-4467-98e2-bbdcb1c574c5.png)

> Additionally, async does not imply Task, it's the type itself that determines how the compiler emits the internals of the asynchronous state machine, so you immediately end up with a circular inference situation.

It is already the behavior in lambda functions:
![image](https://user-images.githubusercontent.com/38350963/190925410-2ec24060-492d-4718-9133-f58cf7436523.png)

There would be some new behavior about `async` functions that I can think out of top of my head:
- Because `Task` would be the default inferred return type of async methods without return statements, `async void` should be written always explicitly.
- Return type of `async` methods containing `yield return` statements should be inferred as `IAsyncEnumerable<ReturnType>`

These are nothing out of ordinary, and I am sure the team can handle that.

But again this is not a detailed proposal because I wanted to check-in first to see if there are other reasons, and performance is the barrier unfortunately according to @CyrusNajmabadi

**Reply by CyrusNajmabadi** [reply]
> But we have already been doing this without scanning the method body, in languages like TS. 

I don't know what you mean. When I use TS that doesn't annotate the return type, I do indeed have to scan to figure it out. 

This is not a problem in TS either as annotations are just an easily ignorable hint to the compiler.  In c#/.net this is not the case. The signature use your contract and it is statically enforced. It's an abi break every time it changes.  JS has no such concept. If I change the TS annotation, nothing at all changes with my program.  All downstream code continues to work exactly the same. 
The JS and .net ecosystems work extremely different in this regard. They cannot be compared in isolation on things like inference due to those differences.

**Reply by CyrusNajmabadi** [reply]
> Having var in C# doesn't make us devs to reanalyze the whole code to understand the variable's types

Variables are not part of your abi. . Changing them affects nothing. That *is* the paradigm of .net.  Things like signatures (including fields) are different and that's why we intentionally believe it's critical to be explicit here.  

Finally, none of the arguments here are changing the two fundamental issues I raised above. First, that we think being explicit about the abi use important. Second, that we couldn't relax this either due to the very real world perf impact of the change.

**Reply by CyrusNajmabadi** [reply]
> But again this is not a detailed proposal because I wanted to check-in first to see if there are other reasons, and performance is the barrier unfortunately according to @CyrusNajmabadi

Perf is one of the barriers. But equally applicable is the abi concern. 

Note, I would be fine with local functions using inference as that would not affect your abi . Toss would also work for top level local functions. 

Here, the inference would be self contained to a single method. Which seems ok to me.

**Reply by CyrusNajmabadi** [reply]
> Because Task would be the default inferred return type of async methods without return statements, async void should be written 

I don't like that at all.  There are many times I would want ValueTask inferred. I don't think we should special case Task in the language.

**Reply by CyrusNajmabadi** [reply]
> to think about the contract of some private helper functions which of course do have their contract, but is not that relevant to the developer.

I would be fine with helper *local* functions using return type inference. This would also apply to top level statements (which just has local functions emitted into the Main method). 

We'd need a separate proposal here though, scoping down to just that feature. @captainsafia might be interested in that as well.

**Reply by CyrusNajmabadi** [reply]
> Having var in C# doesn't make us devs to reanalyze the whole code to understand the variable's types 

Right, but we actually designed it that way. You can just look at the initializer to figure it out. We intentionally do not look further into the body, and we have rejected proposals to do that (e.g. widening the type based on later usage, it allowing the initialization to happen elsewhere)

This is *very* different from return-type inference which now must examine the entire body and understand every return statement and how thry all relate to figure this out. The two cases are not comparable as their differences are precisely why we do this for one and not the other.

**Reply by a11delavar** [reply]
> I don't know what you mean. When I use TS that doesn't annotate the return type, I do indeed have to scan to figure it out.

The screenshots I provided visualized this well!
![image](https://user-images.githubusercontent.com/38350963/190927467-a3174c0c-b17f-4175-b0f1-528ff20fb3e5.png)

> This is not a problem in TS either as annotations are just an easily ignorable hint to the compiler. In c#/.net this is not the case. The signature use your contract and it is statically enforced.

This is only about the idea of inference of the return types and there should be no other implications. TypeScript is just a placeholder which happens to throw the types out even if they are explicitly there. Dotnet would do the opposite and add them back in, in the lowering process, so in the end program IL code there would be 0 change, as it is only suger.

> Variables are not part of your abi. . Changing them affects nothing. That is the paradigm of .net. Things like signatures (including fields) are different and that's why we intentionally believe it's critical to be explicit here.

> Perf is one of the barriers. But equally applicable is the abi concern.

Interesting. I am not familiar with the abi concern, could you kindly elaborate, or point me to some resources?

> I don't like that at all. There are many times I would want ValueTask inferred. I don't think we should special case Task in the language.

Well, my point is, it is already decided upon what the default behavior is for the lambda functions and for anything different you should fall back to explicitly writing the return type yourself. which is absolutely fine.

Even if you wanted to ever make the `Task` assumption for `async` functions configurable, I'm all for it, but don't think it should be in this proposal.

> I would be fine with helper local functions using return type inference. This would also apply to top level statements (which just has local functions emitted into the Main method).

Honestly, it is already the case that I notice people avoiding local functions in favor of lambda functions for the inference I assume.

**Reply by CyrusNajmabadi** [reply]
> The screenshots I provided visualized this well!

The screenshots demonstrate my point. This is a very expensive thing to have to figure out and it *is* a preexisting perf concern and issue for TS. Just because they have a design that has caused them a serious perf concern (among others) does not motivate us to go the same path. It's an anti-motivation.

**Reply by CyrusNajmabadi** [reply]
> This is only about the idea of inference of the return types and there should be no other implications

You cannot just decide this. A return type is part of your abi. It's a contract in .net.  A change there breaks *all* existing compiled callers. This is diametrically different from JS which does not have the concept of these contracts and where the runtime JS code is *entirely* unaffected by the inferred TS types.  

Another way of putting this is that in JS, TS types don't actually exist. So there are no abi risks (it even any abi effect) off anything happens there.  This is the exact opposite of .net where these types *do* exist and any change *at all* will break existing compiled callers.

**Reply by CyrusNajmabadi** [reply]
> Interesting. I am not familiar with the abi concern, could you kindly elaborate, or point me to some resources?

See above, but the basic idea is that. Net is strongly and statically typed. These types are baked into both the API surface areas and into the callsite that uses it. If you change that signature, your consumers immediately break. 

As such, inference is *actively* dangerous as your api/abi may change with *zero* way to tell at the most important location you want to know this.  We consider this to be an *extremely" bad thing, at complete odds with scalable development in .net. 

In JS this doesn't matter as all the TS types are literally erased. They do not actually exist and whatever happens in TS land has *zero* effect on JS land and in your existing callers. 

So the two ecosystems are extremely different with entirely different goals and constraints wrt types and how we think about them. To TS, a type is a hint. Akin to nothing more than a comment in code to help you out. In .net the type is the contract, the cornerstone off what your are delivering, and impossible to ignore. We thus believe the language should mirror that aspect.

**Reply by CyrusNajmabadi** [reply]
> Honestly, it is already the case that I notice people avoiding local functions in favor of lambda functions for the inference I assume.

As before, I'm fine with a scoped proposal for this. It addresses both of the main blockers I've listed above. Namely, there is no abi concern, nor is there a fundamental perf/architecture concern.  It would certainly make method body compilation more expensive if it had inferred local functions, but it would always be scoped to a single method body max. There would never be impact across methods.

**Reply by corruptedlotus** [reply]
I am lost dear Cyrus, how in any way is inference going to interact with ABI? Inference is a high level language mechanism and basically disappears with lowering, which is before IL and Bin. Is it possible that you're mistaking dynamic typing with inference?

**Reply by CyrusNajmabadi** [reply]
> as it is only suger.

To us, this is a bad thing. It hides something critical, and it causes "spooky action at a distance". E.g you change something somewhere you think is innocuous (like the return type of a helper), and now downstream your entire API has changed in a way you didn't realize, and now people who compiled against you are broken. 

Are you going to review your entire surface area after every change to the type of some existing member to validate none of it changed? You will have to for everything that returns 'var'. That can't happen in .net today because you specify your contract, and now it can't ever change without you explicitly choosing you want that.

**Reply by CyrusNajmabadi** [reply]
> how in any way is inference going to interact with ABI? 

```c#
public var ApiMethod()
{
     ...
    return X();
    ...
    return Y();
}
```

This method's return type is inferred from the body. If you change anything this method depends on (like the expressions that affect X() or Y(), which themselves may have inferred return types) then this method's return type changes. 

That's an abi break for any compiled caller, and it cannot be determined without manually going and checking the downstream impact of every change you make. 

This is diametrically different than today where your abi cannot change without an explicit edit at the declaration site of the symbol.** 

We view this as a *bad* thing. 

--
** Note: even NRT preserved this. You could flip the switch on your project and now `string` and `string?` would mean different things. But wrt to the binary abi, they are identical and all existing compiled callers would continue to work. That was not an accident. That fell out entirely because we consider this aspect of the language and compat vital, and we do not compromise on it :-)

**Reply by CyrusNajmabadi** [reply]
> Inference is a high level language mechanism and basically disappears with lowering, which is before IL and Bin. 

This is 100% incorrect. Inference here affects the return type, which is part of the metadata tables. It's not just IL that is affected. That's entirely why we won't do this for methods, but why we would potentially be willing to do this for local functions. 

Being local, they're not part off any contractually exported abi.  So they can change without harming/breaking anyone downstream. That is a critical design philosophy for our language that we care deeply about.

**Reply by HaloFour** [reply]
Perf issues aside I personally wouldn't object to the notion of inference for private members either.

My understanding was that when local functions were being designed that there was a choice as to whether to support inference or hoisting, that the latter would make the former that much more difficult to implement.
